### PR TITLE
check if pod is already running before relaunch

### PR DESCRIPTION
**Pull request by majieyue**
### What changes were proposed in this pull request?

check pod is duplicated before relaunch another one

### Why are the changes needed?

some k8s plugin can recover pod from starting failure. when a pod is deleted or modified from a pending/running state, the pod with the same rank and id may be already running successfully. so we should not relaunch another same rank pod

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

BVT UT

### CODE REVIEW

File: dlrover/python/master/node/dist_job_manager.py [hunk 1]
```diff
@@ -790,6 +790,34 @@ def _process_event(self, event: NodeEvent):
             should_relaunch = self._should_relaunch(
                 cur_node, status_change_flow
             )
+
+            # Even _should_relaunch return True by state-machine check,
+            # we need another round of check
+            if should_relaunch:
```
Comment (workingloong) [inline dlrover/python/master/node/dist_job_manager.py:796]: Move these codes into the method `_should_relaunch` and add UT.

**Review by workingloong — COMMENTED** [review]
Does the native k8s recover the pod from starting failure?

File: dlrover/python/master/node/dist_job_manager.py [hunk 2]
```diff
@@ -790,6 +790,34 @@ def _process_event(self, event: NodeEvent):
...
+                selector = k8s_util.gen_k8s_label_selector_from_dict(
+                    self._get_pod_unique_labels(cur_node)
+                )
+                logger.info(
```
Comment (BalaBalaYi) [inline dlrover/python/master/node/dist_job_manager.py:800]: This part has actually already been implemented (when generating the event before relaunch), so the issue probably isn't here.

**Review by BalaBalaYi — CHANGES_REQUESTED** [review]
need more investigation

File: dlrover/python/master/node/dist_job_manager.py (hunk 1 above)
Comment (majieyue) [reply dlrover/python/master/node/dist_job_manager.py:796]: it is a k8s plugin going to open sourced

File: dlrover/python/master/node/dist_job_manager.py (hunk 2 above)
Comment (majieyue) [reply dlrover/python/master/node/dist_job_manager.py:800]: event type is MODIFIED and node status is RUNNING from log. so the condition in _process_event is not matched.
we didn't see the log either

File: dlrover/python/master/node/dist_job_manager.py (hunk 1 above)
Comment (majieyue) [reply dlrover/python/master/node/dist_job_manager.py:796]: ok

**Comment by majieyue** [comment]
> Does the native k8s recover the pod from starting failure?

it is a k8s plugin going to open sourced

File: dlrover/python/master/node/dist_job_manager.py (hunk 2 above)
Comment (BalaBalaYi) [reply dlrover/python/master/node/dist_job_manager.py:800]: modified event with deletion timestamp will transfer to deleted event:
https://github.com/intelligent-machine-learning/dlrover/blob/5d070f49da6bde5389b0d282a1712ff6cf27d2b8/dlrover/python/master/watcher/k8s_watcher.py#L232

**Comment by BalaBalaYi** [comment]
The issue is resolved by: https://github.com/intelligent-machine-learning/dlrover/pull/1408.

**Pull request closed without merging**